File: smoke_tests/lua/weak_crypto.lua
-- Weak Cryptography vulnerabilities in Lua

local crypto = require("crypto") -- LuaCrypto
local openssl = require("openssl")

-- Test 1: MD5 for password hashing
function hash_password_md5(password)
    -- VULNERABLE: MD5 is broken
    return crypto.digest("md5", password)
end

-- Test 2: SHA1 for password hashing
function hash_password_sha1(password)
    -- VULNERABLE: SHA1 is weak for passwords
    return crypto.digest("sha1", password)
end

-- Test 3: Unsalted hash
function hash_without_salt(password)
    -- VULNERABLE: No salt
    return crypto.digest("sha256", password)
end

-- Test 4: DES encryption
function encrypt_des(data, key)
    -- VULNERABLE: DES is weak
    return crypto.encrypt("des", data, key)
end

-- Test 5: ECB mode
function encrypt_ecb(data, key)
    -- VULNERABLE: ECB mode leaks patterns
    return crypto.encrypt("aes-128-ecb", data, key)
end

-- Test 6: Weak key generation
function generate_weak_key()
    -- VULNERABLE: Predictable key from time
    math.randomseed(os.time())
    local key = ""
    for i = 1, 16 do
        key = key .. string.char(math.random(0, 255))
    end
    return key
end

-- Test 7: RC4 cipher
function encrypt_rc4(data, key)
    -- VULNERABLE: RC4 is broken
    return crypto.encrypt("rc4", data, key)
end

-- Test 8: Static IV
function encrypt_with_static_iv(data, key)
    -- VULNERABLE: Static/predictable IV
    local iv = "0000000000000000"
    return crypto.encrypt("aes-128-cbc", data, key, iv)
end

-- Test 9: Weak random for crypto
function generate_token()
    -- VULNERABLE: math.random is not cryptographically secure
    math.randomseed(os.time())
    local token = ""
    for i = 1, 32 do
        token = token .. string.format("%02x", math.random(0, 255))
    end
    return token
end

-- Test 10: Password comparison timing attack
function verify_password(input, stored)
    -- VULNERABLE: Non-constant time comparison
    return input == stored
end

-- Test 11: Weak PBKDF iterations
function derive_key(password, salt)
    -- VULNERABLE: Too few iterations
    return crypto.pbkdf2("sha256", password, salt, 100, 32)
end

-- Test 12: Hardcoded salt
function hash_with_hardcoded_salt(password)
    -- VULNERABLE: Same salt for all users
    local salt = "static_salt_value"
    return crypto.digest("sha256", password .. salt)
end

-- Test 13: Blowfish with small key
function encrypt_blowfish(data, key)
    -- VULNERABLE: Blowfish with weak key
    if #key < 16 then
        -- Using short key
    end
    return crypto.encrypt("bf", data, key)
end

-- Test 14: Base64 as "encryption"
function pseudo_encrypt(data)
    -- VULNERABLE: Base64 is not encryption
    local mime = require("mime")
    return mime.b64(data)
end

-- Test 15: RSA with small key
function rsa_encrypt_weak(data)
    -- VULNERABLE: RSA key too small
    local rsa = openssl.rsa.generate(512)  -- Should be 2048+
    return rsa:encrypt(data)
end

-- Test 16: SHA256 for HMAC without proper key
function weak_hmac(data)
    -- VULNERABLE: Weak/short HMAC key
    return crypto.hmac("sha256", data, "short")
end

File: smoke_tests/perl/weak_crypto.pl
#!/usr/bin/perl
# Weak Cryptography vulnerabilities in Perl

use strict;
use warnings;
use Digest::MD5;
use Digest::SHA;
use Crypt::CBC;
use Crypt::DES;

# Test 1: MD5 for password hashing
sub hash_password_md5 {
    my $password = shift;
    # VULNERABLE: MD5 is weak for passwords
    my $hash = Digest::MD5::md5_hex($password);
    return $hash;
}

# Test 2: SHA1 for security purposes
sub hash_password_sha1 {
    my $password = shift;
    # VULNERABLE: SHA1 is deprecated for security
    my $hash = Digest::SHA::sha1_hex($password);
    return $hash;
}

# Test 3: DES encryption
sub encrypt_des {
    my ($plaintext, $key) = @_;
    # VULNERABLE: DES is weak
    my $cipher = Crypt::CBC->new(
        -key    => $key,
        -cipher => 'DES'
    );
    return $cipher->encrypt($plaintext);
}

# Test 4: ECB mode
sub encrypt_ecb {
    my ($plaintext, $key) = @_;
    # VULNERABLE: ECB mode leaks patterns
    my $cipher = Crypt::CBC->new(
        -key    => $key,
        -cipher => 'Blowfish',
        -header => 'none',
        -literal_key => 1,
        -keysize => 16,
        -padding => 'null',
    );
    # Using raw mode (effectively ECB)
    return $cipher->encrypt($plaintext);
}

# Test 5: Weak key derivation
sub derive_key_weak {
    my $password = shift;
    # VULNERABLE: Simple MD5 for key derivation
    return Digest::MD5::md5($password);
}

# Test 6: Hardcoded IV
sub encrypt_with_fixed_iv {
    my ($plaintext, $key) = @_;
    # VULNERABLE: Static IV
    my $iv = "1234567890123456";
    my $cipher = Crypt::CBC->new(
        -key    => $key,
        -cipher => 'Rijndael',
        -iv     => $iv,
        -header => 'none',
    );
    return $cipher->encrypt($plaintext);
}

# Test 7: Insecure random for crypto
sub generate_weak_token {
    # VULNERABLE: rand() is not cryptographically secure
    my $token = "";
    for (1..32) {
        $token .= sprintf("%02x", int(rand(256)));
    }
    return $token;
}

# Test 8: RC4 encryption
sub encrypt_rc4 {
    my ($plaintext, $key) = @_;
    # VULNERABLE: RC4 is broken
    use Crypt::RC4;
    my $rc4 = Crypt::RC4->new($key);
    return $rc4->RC4($plaintext);
}

# Test 9: Short key length
sub encrypt_short_key {
    my $plaintext = shift;
    # VULNERABLE: Key too short
    my $key = "short";
    my $cipher = Crypt::CBC->new(
        -key    => $key,
        -cipher => 'Rijndael',
    );
    return $cipher->encrypt($plaintext);
}

# Test 10: Blowfish with weak key
sub encrypt_blowfish_weak {
    my ($plaintext, $password) = @_;
    # VULNERABLE: Using password directly as key
    my $cipher = Crypt::CBC->new(
        -key    => $password,
        -cipher => 'Blowfish',
    );
    return $cipher->encrypt($plaintext);
}

# Test 11: CRC32 for integrity
sub compute_checksum {
    use String::CRC32;
    my $data = shift;
    # VULNERABLE: CRC32 is not secure for integrity
    return crc32($data);
}

# Test 12: Base64 as "encryption"
sub encode_secret {
    use MIME::Base64;
    my $secret = shift;
    # VULNERABLE: Base64 is encoding, not encryption
    return encode_base64($secret);
}

# Test 13: XOR cipher
sub xor_encrypt {
    my ($plaintext, $key) = @_;
    # VULNERABLE: XOR cipher is easily broken
    my $encrypted = "";
    my @key_chars = split //, $key;
    my $i = 0;
    for my $char (split //, $plaintext) {
        $encrypted .= chr(ord($char) ^ ord($key_chars[$i % @key_chars]));
        $i++;
    }
    return $encrypted;
}

# Test 14: Predictable salt
sub hash_with_predictable_salt {
    my $password = shift;
    # VULNERABLE: Predictable salt
    my $salt = "fixed_salt_value";
    return Digest::SHA::sha256_hex($salt . $password);
}

# Test 15: No salt for password hash
sub hash_without_salt {
    my $password = shift;
    # VULNERABLE: No salt
    return Digest::SHA::sha256_hex($password);
}

# Test 16: Weak PRNG seed
sub seed_random {
    # VULNERABLE: Weak seeding
    srand(time());
    return int(rand(1000000));
}

1;
